src/simulators/ch/ch_state.hpp

```diff
@@ -0,0 +1,787 @@
+
+  virtual std::vector<reg_t> sample_measure(const reg_t& qubits,
+                                            uint_t shots,
+                                            RngEngine &rng) override;
+
+protected:
+
+  //Alongside the sample measure optimisaiton, we can parallelise
+  //circuit applicaiton over the states. This reduces the threading overhead
+  //as we only have to fork once per circuit.
+  void apply_ops_parallel(const std::vector<Operations::Op> &ops,
+                                  RngEngine &rng);
+
+  //Small routine that eschews any parallelisation/decomposition and applies a stabilizer
+  //circuit to a single state. This is used to optimize a circuit with a large
+  //initial clifford fraction, or for running stabilizer circuits.
+  void apply_stabilizer_circuit(const std::vector<Operations::Op> &ops,
+                                      OutputData &data,
+                                      RngEngine &rng);
+  // Applies a sypported Gate operation to the state class.
+  // If the input is not in allowed_gates an exeption will be raised.
+  // TODO: Investigate OMP synchronisation over stattes to remove these different versions
+  // One option would be tasks, but the memory overhead isn't clear
+  void apply_gate(const Operations::Op &op, RngEngine &rng);
+  void apply_gate(const Operations::Op &op, RngEngine &rng, uint_t rank);
+
+  // Measure qubits and return a list of outcomes [q0, q1, ...]
+  // If a state subclass supports this function then "measure" 
+  // should be contained in the set returned by the 'allowed_ops'
+  // method.
+  void apply_measure(const reg_t &qubits,
+                     const reg_t &cmemory,
+                     const reg_t &cregister,
+                     RngEngine &rng);
+
+  // Reset the specified qubits to the |0> state by measuring the
+  // projectors Id+Z_{i} for each qubit i
+  void apply_reset(const reg_t &qubits, AER::RngEngine &rng);
+
+  //Take a snapshot of the simulation state
+  //TODO: Improve the CHSimulator::to_json method.
+  void apply_snapshot(const Operations::Op &op, OutputData &data, RngEngine &rng);
+  //Convert a decomposition to a state-vector
+  void statevector_snapshot(const Operations::Op &op, OutputData &data, RngEngine &rng);
+  //Compute probabilities from a stabilizer rank decomposition
+  //TODO: Check ordering/output format...
+  void probabilities_snapshot(const Operations::Op &op, OutputData &data, RngEngine &rng);
+
+  const static stringmap_t<Gates> gateset_;
+  const static stringmap_t<Snapshots> snapshotset_;
+
+  //-----------------------------------------------------------------------
+  //Parameters and methods specific to the Stabilizer Rank Decomposition
+  //-----------------------------------------------------------------------
+
+  //Allowed error in the stabilizer rank decomposition.
+  //The required number of states scales as \delta^{-2}
+  //for allowed error \delta
+  double approximation_error = 0.05;
```

Comment:
Private class variables should be suffixed with underscore: `approximation_error_ = 0.05`